### PR TITLE
Implement JSON Merge Patch  (RFC 7396 #623)

### LINKED ISSUE #623: Implement RFC 7396 (json merge)
original: https://github.com/mozilla-services/cliquet/issues/597

The proposal is:  If None is set in a PATCH request for a record attribute, that removes it

### PR DESCRIPTION
**Pull request by gabisurita**
Fixes #623 

- [x] Add tests.
- [x] Add a changelog entry.
- [x] If you changed the HTTP API, update the [API_VERSION](https://github.com/Kinto/kinto/blob/master/kinto/__init__.py#L15) constant and add an API changelog entry [in the docs](https://github.com/Kinto/kinto/blob/master/docs/api/index.rst)

Let me know if I got it wrong or there's something missing! 

Obs: Should this increase API_VERSION or be included on the API entry docs?

### CODE REVIEW

**Comment by gabisurita** [comment]
Behaviour:

```
$ echo '{"data" : {"touched": "yes" }}' | http PATCH http://localhost:8888/v1/buckets/default/collections/c0/records/r0 --auth aaa:aaa
HTTP/1.1 200 OK
Access-Control-Expose-Headers: Retry-After, Content-Length, Alert, Backoff
Content-Length: 169
Content-Type: application/json; charset=UTF-8
Date: Mon, 10 Oct 2016 21:15:54 GMT
Etag: "1476134154134"
Last-Modified: Mon, 10 Oct 2016 21:15:54 GMT
Server: waitress

{
    "data": {
        "id": "r0", 
        "last_modified": 1476134154134, 
        "touched": "yes"
    }, 
    "permissions": {
        "write": [
            "basicauth:b4a6199b0cbf1363ef8e60c1c26e0358f49f5698c6bb4cb879437732768fcc94"
        ]
    }
}

$ echo '{"data" : {"touched": null }}' | http PATCH http://localhost:8888/v1/buckets/default/collections/c0/records/r0 --auth aaa:aaa
HTTP/1.1 200 OK
Access-Control-Expose-Headers: Retry-After, Content-Length, Alert, Backoff
Content-Length: 153
Content-Type: application/json; charset=UTF-8
Date: Mon, 10 Oct 2016 21:16:09 GMT
Etag: "1476134169802"
Last-Modified: Mon, 10 Oct 2016 21:16:09 GMT
Server: waitress

{
    "data": {
        "id": "r0", 
        "last_modified": 1476134169802
    }, 
    "permissions": {
        "write": [
            "basicauth:b4a6199b0cbf1363ef8e60c1c26e0358f49f5698c6bb4cb879437732768fcc94"
        ]
    }
}
```

File: kinto/core/resource/__init__.py [hunk 1]
```diff
@@ -474,6 +474,9 @@ def patch(self):
 
         updated = self.apply_changes(existing, changes=changes)
 
+        # remove field if null attribute is passed (RFC 7396)
+        updated = {key: value for key, value in updated.items() if value}
```
Comment (Natim) [inline kinto/core/resource/__init__.py]: I would use `if value is not None` because we don't want to remove it for `False` and `0`
You might to add a test to make sure we don't remove the field with `False`, `0`, `""` (empty string) etc.

**Comment by Natim** [comment]
That's so great thanks !
I think the API doc must be updated as well :)

**Comment by leplatrem** [comment]
> Let me know if I got it wrong or there's something missing!

Nothing wrong! But yes some stuff is missing [from the spec](https://tools.ietf.org/html/rfc7396)

* Only use this behaviour when the specific `Content-Type` is used
* Merge recursively the objects (see a pseudo code example is given in the RFC)

I believe you should move this to `core.utils` and run some of the tests specified in the specs (especially the last one).

I know it would be better if we would have this new behaviour by default, but we can't break existing clients, so we better leave behind the request header `Content-Type: application/merge-patch+json` as the spec says.

> Obs: Should this increase API_VERSION or be included on the API entry docs?

Yes it should be incremented!

File: tests/core/resource/test_record.py [hunk 2]
```diff
@@ -227,6 +227,15 @@ def test_patch_record_returns_updated_fields(self):
         self.assertEquals(self.stored['id'], self.result['id'])
         self.assertEquals(self.result['position'], 10)
 
+    def test_patch_removes_attribute_if_none(self):
+        self.resource.request.json = {'data': {'field': 'aaa'}}
+        self.resource.patch()['data']
+        self.resource.request.json = {'data': {'field': None}}
+        result = self.resource.patch()['data']
+        self.assertEquals('field' in result.keys(), False)
```
Comment (leplatrem) [inline tests/core/resource/test_record.py]: You can use `self.assertNotIn('field', result)`

File: kinto/core/resource/__init__.py (hunk 1 above)
Comment (leplatrem) [inline kinto/core/resource/__init__.py]: Theses should be moved to `self.apply_changes()` I think

**Comment by gabisurita** [comment]
> I know it would be better if we would have this new behaviour by default, but we can't break existing clients, so we better leave behind the request header Content-Type: application/merge-patch+json as the spec says.

That would include accept 'application/merge-patch+json' as a valid content type on the viewset, right?

Running the RFC example over a record: 

```
$ echo '{"data":{ "title": "Goodbye!", "author" : { "givenName" : "John", "familyName" : "Doe" }, "tags":[ "example", "sample" ], "content": "This will be unchanged" }}' | http PATCH http://localhost:8888/v1/buckets/default/collections/c0/records/r1 --auth aaa:aaa
HTTP/1.1 200 OK
Access-Control-Expose-Headers: Retry-After, Content-Length, Alert, Backoff
Content-Length: 317
Content-Type: application/json; charset=UTF-8
Date: Tue, 11 Oct 2016 15:52:53 GMT
Etag: "1476201173217"
Last-Modified: Tue, 11 Oct 2016 15:52:53 GMT
Server: waitress

{
    "data": {
        "author": {
            "familyName": "Doe", 
            "givenName": "John"
        }, 
        "content": "This will be unchanged", 
        "id": "r1", 
        "last_modified": 1476201173217, 
        "phoneNumber": "+01-123-456-7890", 
        "tags": [
            "example", 
            "sample"
        ], 
        "title": "Goodbye!"
    }, 
    "permissions": {
        "write": [
            "basicauth:b4a6199b0cbf1363ef8e60c1c26e0358f49f5698c6bb4cb879437732768fcc94"
        ]
    }
}

[castro@localhost kinto]$ echo '{"data":{ "title": "Hello!", "phoneNumber": "+01-123-456-7890", "author": { "familyName": null }, "tags": [ "example" ] }}' | http PATCH http://localhost:8888/v1/buckets/default/collections/c0/records/r1 Content-Type:application/merge-patch+json --auth aaa:aaa
HTTP/1.1 200 OK
Access-Control-Expose-Headers: Retry-After, Content-Length, Alert, Backoff
Content-Length: 287
Content-Type: application/json; charset=UTF-8
Date: Tue, 11 Oct 2016 15:52:54 GMT
Etag: "1476201174716"
Last-Modified: Tue, 11 Oct 2016 15:52:54 GMT
Server: waitress

{
    "data": {
        "author": {
            "givenName": "John"
        }, 
        "content": "This will be unchanged", 
        "id": "r1", 
        "last_modified": 1476201174716, 
        "phoneNumber": "+01-123-456-7890", 
        "tags": [
            "example"
        ], 
        "title": "Hello!"
    }, 
    "permissions": {
        "write": [
            "basicauth:b4a6199b0cbf1363ef8e60c1c26e0358f49f5698c6bb4cb879437732768fcc94"
        ]
    }
}
```

**Comment by leplatrem** [comment]
This is just excellent! Awesome job, definitely not easy!

We'll probably put that header by default in our official python/js clients.

**Comment by gabisurita** [comment]
@leplatrem 

I just realized there is a problem with permissions related to the different content type header. For some reason, permissions is not  validated on [process_record](https://github.com/Kinto/kinto/blob/master/kinto/core/resource/__init__.py#L1142).

A quick fix would be get permissions not from ``request.validated``, but directly from ``request.json``, but I sure if this breaks something else.

Any clue why this is happening? Should we open an issue?

**Comment by gabisurita** [comment]
Edit: never mind. Everything is ok, I've just used a badly formatted request.
